### PR TITLE
added initial assignment

### PR DESCRIPTION
**Pull request by girishg4t**
Done the assignment please review

### CODE REVIEW

File: deployment.yaml [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+spec:
+  replicas: 3
+  selector:
+    matchLabels:
+      app: nginx
+  template:
+    metadata:
+      annotations:
+        auto-create-svc: "true"
+        auto-create-svc-type: "NodePort"
```
Comment (viveksinghggits) [inline deployment.yaml]: I think the requirement is to have these annotations set for the deployment resource itself and not the pod resource.

File: main.go [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+
+var api *k8sServiceAPI
+
+func main() {
+	api = &k8sServiceAPI{
+		cs: getKubeHandle(),
+	}
+
+	watchlist := cache.NewListWatchFromClient(api.cs.AppsV1().RESTClient(), "deployments",
+		ns, fields.Everything())
```
Comment (viveksinghggits) [inline main.go]: lets try to watch the deployment resource using clientset only, instead of RESTClient

File: main.go [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+		time.Sleep(time.Second)
+	}
+}
+
+func onAdd(obj interface{}) {
+	dep := obj.(*v1.Deployment)
+	name := dep.GetName()
+
+	if name == "" {
+		return
```
Comment (viveksinghggits) [inline main.go]: let's try to return the message along with the return statement.

File: main.go [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+	}
+	service, err := extractServiceInfoFromDeployment(dep)
+	if err != nil {
+		return
+	}
+	api.s = service
+	exists := isServiceExists(api)
```
Comment (viveksinghggits) [inline main.go]: let's try to make this mode readable, by making a controller struct and have client fields in that controller struct.

File: main.go [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+	if exists {
+		return
+	}
+	err = createService(api)
+	if err != nil {
+		panic(err)
```
Comment (viveksinghggits) [inline main.go]: let's not panic here, return the error and handle in the main.

File: service.go [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+}
+
+func extractServiceInfoFromDeployment(dep *v1.Deployment) (*serviceResource, error) {
+	metadata := dep.Spec.Template.ObjectMeta
+	spec := dep.Spec.Template.Spec
+	serviceName := dep.GetName() + "-service"
+	isServiceReq, _ := strconv.ParseBool(metadata.Annotations["auto-create-svc"])
+	if !isServiceReq {
+		return nil, errors.New("service not required")
+	}
```
Comment (viveksinghggits) [inline service.go]: we can move this check to, before calling the function.

**Review by viveksinghggits — CHANGES_REQUESTED** [review]

File: deployment.yaml (hunk 1 above)
Comment (girishg4t) [reply deployment.yaml]: done

File: main.go (hunk 2 above)
Comment (girishg4t) [reply main.go]: done

File: main.go (hunk 3 above)
Comment (girishg4t) [reply main.go]: This are default event handlers and we can't return any error or message from that , lets discuss

File: main.go (hunk 4 above)
Comment (girishg4t) [reply main.go]: new controller added

File: service.go (hunk 6 above)
Comment (girishg4t) [reply service.go]: done

File: main.go (hunk 5 above)
Comment (girishg4t) [reply main.go]: This are default event handlers and we can't return any error or message from that , lets discuss

**Pull request merged**